godot-core/src/builtin/meta/godot_convert/convert_error.rs

```diff
@@ -0,0 +1,260 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::error::Error;
+use std::fmt;
+
+use godot_ffi::VariantType;
+
+use crate::builtin::{array_inner, meta::ClassName};
+
+type Cause = Box<dyn Error + Send + Sync>;
+
+/// Represents errors that can occur when converting values from Godot.
+#[derive(Debug)]
+pub struct ConvertError {
+    kind: ErrorKind,
+    cause: Option<Cause>,
+    value: Option<Box<dyn fmt::Debug>>,
+}
+
+impl ConvertError {
+    /// Create a new custom error for a conversion.
+    pub fn new() -> Self {
+        Self {
+            kind: ErrorKind::Custom,
+            cause: None,
+            value: None,
+        }
+    }
+
+    fn custom() -> Self {
+        Self {
+            kind: ErrorKind::Custom,
+            cause: None,
+            value: None,
+        }
+    }
+
+    /// Create a new custom error for a conversion with the value that failed to convert.
+    pub fn new_with_value<V: fmt::Debug + 'static>(value: V) -> Self {
+        let mut err = Self::custom();
+        err.value = Some(Box::new(value));
+        err
+    }
+
+    /// Create a new custom error with a rust-error as an underlying cause for the conversion error.
+    pub fn new_with_cause<C: Into<Cause>>(cause: C) -> Self {
+        let mut err = Self::custom();
+        err.cause = Some(cause.into());
+        err
+    }
+
+    /// Create a new custom error with a rust-error as an underlying cause for the conversion error, and the
+    /// value that failed to convert.
+    pub fn new_with_cause_value<C: Into<Cause>, V: fmt::Debug + 'static>(
+        cause: C,
+        value: V,
+    ) -> Self {
+        let mut err = Self::custom();
+        err.cause = Some(cause.into());
+        err.value = Some(Box::new(value));
+        err
+    }
```

Comment:
Nitpick, maybe `with_*` instead of `new_with_*` would be enough?

Interestingly, Rust itself has both 🙈 `Vec::with_capacity` and `SipHasher::new_with_keys`. Although the former is much more common and not deprecated...